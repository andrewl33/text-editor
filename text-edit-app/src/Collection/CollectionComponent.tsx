import * as React from "react";
import { Container, Grid } from "semantic-ui-react";
import { ListComponent } from "../generic/Lists/ListComponent";
import SidebarComponent from "../generic/Sidebar/SidebarComponent";
import { CollectionComponentProps } from "../types";
// TODO: add styling
export const CollectionComponent = (props: CollectionComponentProps) => {
  const { name, createDate, isLocked, users } = props;

  return (
    <Grid style={{ marginTop: 0, height: "100%" }}>
      <Grid.Column width={13} style={{ paddingRight: 0, paddingTop: 0 }}>
        <ListComponent header="Files" {...props} />
      </Grid.Column>
      <Grid.Column
        width={3}
        textAlign="center"
        style={{
          paddingRight: "2em",
          backgroundColor: "#3180d4",
          height: "100%"
        }}
      >
        <Container>
          <SidebarComponent
            name={name}
            pageType="collection"
            createDate={createDate}
            isPrivate={isLocked}
            users={users}
            // tslint:disable-next-line
            onNameChange={a => {}}
          />
        </Container>
      </Grid.Column>
    </Grid>
  );
};
