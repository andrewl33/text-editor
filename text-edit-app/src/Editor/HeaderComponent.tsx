import * as React from 'react';
import { Button, Menu } from 'semantic-ui-react';
import { EditorProps } from '../types';
import { RefObject } from 'react-sane-contenteditable';
import { ROOT_URL } from '../envConstants';
export interface HeaderProps {
  onLock?: () => void;
  onNew?: () => void;
  onShare?: () => void;
  pathname: string;
}


export class HeaderComponent extends React.Component<EditorProps, {copy: boolean}> {
  
  private urlHolder: RefObject<any>;
  
  constructor(props: EditorProps) {
    super(props);

    this.urlHolder = React.createRef();
    this.copyToClipboard = this.copyToClipboard.bind(this);

    this.state = {
      copy: false
    }
  }

  public render() {

    const { onNew, onLock, pathname } = this.props;

    return (
      <header>
        <Menu attached={true} inverted={true} primary="true" size='huge'>
          <Menu.Item name='TextEdit' />
          <Menu.Menu position='right'>
            <Menu.Item>
              <Button primary={true} onClick={onNew}>New</Button>
            </Menu.Item>
            <Menu.Item>
              <Button primary={true} onClick={onLock}>Lock</Button>
            </Menu.Item>
            <Menu.Item>
              <Button primary={true} onClick={this.copyToClipboard}>Share</Button>
            </Menu.Item>
          </Menu.Menu>
        </Menu>
        {this.state.copy ? 
          (<textarea className="url-holder" value={ROOT_URL+pathname} placeholder={"placeholder"} readOnly={true} ref={this.urlHolder}/>)
        : ''}
        
      </header>
    );
  }

  public componentDidUpdate() {
    // copy to clipboard
    // reset copy state
    if (this.state.copy) {
      this.urlHolder.current.select();
      window.document.execCommand('copy');
      this.setState({copy: false});
      this.props.onShare();
    }
  }


  
  private copyToClipboard(): void {
    // shows the textarea so it is selectable
    this.setState({copy: true});

    // the state is set back to false in componentDidUpdate()
  }
}

export default HeaderComponent;