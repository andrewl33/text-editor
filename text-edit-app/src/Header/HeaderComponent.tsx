import * as React from 'react';
import { Button, Menu } from 'semantic-ui-react';

export interface HeaderProps {
  onLock?: () => void;
  onNew?: () => void;
  onShare?: () => void;
}

const TopHeader = ({ onNew, onLock, onShare }: HeaderProps) => {
  return (
    <header>
      <Menu attached={true} inverted={true} primary="true" size='huge'>
        <Menu.Item name='TextEdit' />
        <Menu.Menu position='right'>
          <Menu.Item>
            <Button primary={true} onClick={onNew}>New</Button>
          </Menu.Item>
          <Menu.Item>
            <Button primary={true} onClick={onLock}>Lock</Button>
          </Menu.Item>
          <Menu.Item>
            <Button primary={true} onClick={onShare}>Share</Button>
          </Menu.Item>
        </Menu.Menu>
      </Menu>
    </header>
  );
}

export default TopHeader;