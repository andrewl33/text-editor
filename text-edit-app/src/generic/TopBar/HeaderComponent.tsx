import * as React from "react";
import { Button, Dropdown, Menu, Transition } from "semantic-ui-react";
import { ROOT_URL } from "../../envConstants";
import { HeaderProps } from "../../types";
import AlertComponent from "../Alert/AlertComponent";
import { PromptComponent } from "../Prompt/PromptComponent";

interface HeaderState {
  copy: boolean;
  onLockPrompt: boolean;
}

export class HeaderComponent extends React.Component<HeaderProps, HeaderState> {
  private urlHolder: React.RefObject<HTMLTextAreaElement>;
  private notificationDuration: number;

  constructor(props: HeaderProps) {
    super(props);

    this.urlHolder = React.createRef();
    this.copyToClipboard = this.copyToClipboard.bind(this);

    this.state = {
      copy: false,
      onLockPrompt: false
    };
  }

  public render() {
    const {
      onNew,
      accountName,
      loggedIn,
      pathname,
      pageName,
      isShareable = false,
      openAlert,
      alertMessage,
      getAccountCredentials,
      getPassword,
      prompt,
      onPrompt,
      onDashboard,
      onLogOut,
      onLogInPrompt,
      onClosePrompt,
      onLock,
      onHomeClick
    } = this.props;

    const alert = alertMessage !== undefined && (
      <AlertComponent open={openAlert} message={alertMessage} />
    );
    const shareButton = isShareable && (
      <Menu.Item>
        <Button primary={true} onClick={this.copyToClipboard}>
          Share
        </Button>
      </Menu.Item>
    );
    const hiddenShare = this.state.copy && (
      <textarea
        className="url-holder"
        value={ROOT_URL + pathname}
        placeholder={"placeholder"}
        readOnly={true}
        ref={this.urlHolder}
      />
    );
    const newButton = onNew && (
      <Menu.Item>
        <Button primary={true} onClick={onNew}>
          New
        </Button>
      </Menu.Item>
    );
    const lockButton = onLock && (
      <Menu.Item>
        <Button primary={true} onClick={this.handleLockRequest}>
          Lock
        </Button>
      </Menu.Item>
    );
    const accountDropDown = (
      <div>
        <Dropdown
          text={loggedIn ? accountName : "Account"}
          icon="user"
          floating={true}
          labeled={true}
          button={true}
          className="icon"
          style={{
            color: "#fff",
            backgroundColor: "#2185d0",
            fontSize: 17
          }}
        >
          {loggedIn ? (
            <Dropdown.Menu>
              <Dropdown.Item onClick={onDashboard}>Dashboard</Dropdown.Item>
              <Dropdown.Item onClick={onLogOut}>Log out</Dropdown.Item>
            </Dropdown.Menu>
          ) : (
            <Dropdown.Menu>
              <Dropdown.Item onClick={onLogInPrompt}>Log in</Dropdown.Item>
            </Dropdown.Menu>
          )}
        </Dropdown>
      </div>
    );

    return (
      <header>
        {alert}
        <Menu attached={true} inverted={true} primary="true" size="huge">
          <Menu.Item name="TextEdit" onClick={onHomeClick} />
          <Menu.Item name={pageName} />
          <Menu.Menu position="right">
            <Menu.Item>
              <Button primary={true} onClick={this.props.toGradingPage}>
                Grading
              </Button>
            </Menu.Item>
            {newButton}
            {lockButton}
            {shareButton}
            <Menu.Item>{accountDropDown}</Menu.Item>
          </Menu.Menu>
        </Menu>
        <Transition
          animation="fade down"
          unmountOnHide={true}
          visible={onPrompt || this.state.onLockPrompt}
        >
          <PromptComponent
            getAccountCredentials={getAccountCredentials}
            getPassword={getPassword}
            prompt={this.state.onLockPrompt ? "Lock" : prompt}
            onClosePrompt={onClosePrompt}
            lockItem={this.handleLockSubmit}
          />
        </Transition>
        {hiddenShare}
      </header>
    );
  }

  public componentDidUpdate = (): void => {
    // copy to clipboard
    // reset copy state
    // if (this.props.isShareable && this.state.copy) {
    if (true && this.state.copy) {
      this.urlHolder.current!.select();
      window.document.execCommand("copy");
      this.setState({ copy: false });
    }

    if (this.props.openAlert) {
      clearTimeout(this.notificationDuration);
      this.notificationDuration = window.setTimeout(this.props.onAlert, 1500);
    }
  };

  private copyToClipboard = (): void => {
    // shows the textarea so it is selectable
    this.setState({ copy: true });
    // the state is set back to false in componentDidUpdate()
  };

  private handleLockRequest = (): void => {
    this.setState({ onLockPrompt: true });
  };

  private handleLockSubmit = (password: string) => {
    this.setState({ onLockPrompt: false });
    if (this.props.onLock) {
      this.props.onLock(password);
    }
  };
}
