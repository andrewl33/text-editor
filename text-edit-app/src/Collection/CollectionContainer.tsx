import { push, RouterAction } from "connected-react-router";
import * as React from "react";
import { connect } from "react-redux";
import { ThunkDispatch } from "redux-thunk";

import {
  AuthAction,
  closePrompt,
  logIn,
  logInPrompt,
  logOut
} from "../Auth/AuthAction";
import { HeaderComponent } from "../generic/TopBar/HeaderComponent";
import { CollectionContainerProps, HeaderProps, StoreState } from "../types";
import {
  addFileToCollection,
  addUserToCollection,
  authCollection,
  changeCollectionName,
  closeAlert,
  CollectionAction,
  getCollectionFiles,
  lockCollection,
  newCollection,
  removeFileFromCollection,
  removeUserFromCollection,
  shareLink
} from "./CollectionAction";
import { CollectionComponent } from "./CollectionComponent";

export class CollectionContainer extends React.Component<
  CollectionContainerProps & HeaderProps
> {
  public render() {
    const {
      accountName,
      loggedIn,
      openAlert,
      alertMessage,
      authPrompt,
      pathname,
      onNew,
      onLock,
      onShare,
      onAlert,
      name,
      createDate,
      users,
      isLocked,
      collectionPrompt,
      onNameChange,
      onDashboard,
      onLogInPrompt,
      onLogOut,
      onAddFile,
      onRemoveFile,
      onAddUser,
      onRemoveUser,
      onAuthCollection,
      onAuthAccount,
      onClosePrompt
    } = this.props;

    const files = [
      {
        id: "1",
        name: "Hello World",
        tags: ["Snippet", "Rust", "Mission Critical"],
        date: "1-1-1"
      },
      {
        id: "2",
        name: "test",
        tags: ["css", "firefox", "mobile"],
        date: "1-1-1"
      },
      {
        id: "3",
        name: "",
        tags: ["test", "Google", "jimmy"],
        date: "1-1-1"
      }
    ];

    return (
      <div style={{ height: "100%" }}>
        <HeaderComponent
          onNew={onNew}
          onLock={onLock}
          onShare={onShare}
          onAlert={onAlert}
          accountName={accountName}
          loggedIn={loggedIn}
          pathname={pathname}
          isShareable={true}
          openAlert={openAlert}
          alertMessage={alertMessage}
          pageName={"Collection"}
          prompt={authPrompt ? "Login" : "Private File"}
          getAccountCredentials={onAuthAccount}
          getPassword={onAuthCollection}
          onPrompt={authPrompt || collectionPrompt}
          onLogInPrompt={onLogInPrompt}
          onDashboard={onDashboard}
          onLogOut={onLogOut}
          onClosePrompt={onClosePrompt}
        />
        <CollectionComponent
          items={files}
          createDate={createDate}
          name={name}
          users={users}
          isLocked={isLocked}
          onNameChange={onNameChange}
          onAddFile={onAddFile}
          onRemoveFile={onRemoveFile}
          onAddUser={onAddUser}
          onRemoveUser={onRemoveUser}
        />
      </div>
    );
  }

  public componentDidMount() {
    this.props.onMount();
  }
}

const mapStateToProps = (state: StoreState) => {
  const {
    openAlert,
    alertMessage,
    items,
    name,
    createDate,
    users,
    isLocked,
    collectionPrompt
  } = state.collection;
  const { accountName, loggedIn, authPrompt } = state.authentication;
  const { pathname } = state.router.location;

  return {
    accountName,
    loggedIn,
    openAlert,
    authPrompt,
    alertMessage,
    items,
    name,
    createDate,
    users,
    pathname,
    isLocked,
    collectionPrompt
  };
};

const mapDispatchToProps = (
  dispatch: ThunkDispatch<
    StoreState,
    void,
    AuthAction | CollectionAction | RouterAction
  >
) => {
  return {
    onNew: () => dispatch(newCollection()),
    onLock: (password: string) => dispatch(lockCollection(password)),
    onShare: () => dispatch(shareLink()),
    onAlert: () => dispatch(closeAlert()),
    onMount: () => dispatch(getCollectionFiles()),
    onAddFile: (fileId: string) => dispatch(addFileToCollection(fileId)),
    onRemoveFile: (fileId: string) =>
      dispatch(removeFileFromCollection(fileId)),
    onAddUser: (accountName: string) =>
      dispatch(addUserToCollection(accountName)),
    onRemoveUser: (accountName: string) =>
      dispatch(removeUserFromCollection(accountName)),
    onAuthCollection: (password: string) => dispatch(authCollection(password)),
    onAuthAccount: (name: string, pass: string) => dispatch(logIn(name, pass)),
    // TODO: onAuthCollection: (pass: string) => dispatch(authCollection(pass)),
    onNameChange: (name: string) => dispatch(changeCollectionName(name)),
    onDashboard: () => dispatch(push("/dashboard")),
    onLogInPrompt: () => dispatch(logInPrompt()),
    onLogOut: () => dispatch(logOut()),
    onClosePrompt: () => dispatch(closePrompt())
  };
  // TODO: auth
  // TODO: delete
  // TODO: addfile
  // TODO: remove file
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CollectionContainer);
