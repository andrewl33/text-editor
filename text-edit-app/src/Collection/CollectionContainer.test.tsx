import { shallow } from "enzyme";
import * as React from "react";
import * as renderer from "react-test-renderer";
import { HeaderComponent } from "../generic/TopBar/HeaderComponent";
import { CollectionComponent } from "./CollectionComponent";
import { CollectionContainer } from "./CollectionContainer";

const dispatchProps = {
  onNew: jest.fn(),
  onLock: jest.fn(),
  onShare: jest.fn(),
  onAlert: jest.fn(),
  onMount: jest.fn(),
  onAddFile: jest.fn(),
  onRemoveFile: jest.fn(),
  onAddUser: jest.fn(),
  onRemoveUser: jest.fn(),
  onAuthCollection: jest.fn(),
  onAuthAccount: jest.fn(),
  onNameChange: jest.fn(),
  onDashboard: jest.fn(),
  onLogInPrompt: jest.fn(),
  onLogOut: jest.fn(),
  onClosePrompt: jest.fn(),
  onFileClick: jest.fn(),
  onHomeClick: jest.fn()
};

const colProps = {
  openAlert: false,
  alertMessage: "None",
  items: [],
  name: "name",
  createDate: "1-1-2001",
  users: [],
  isLocked: false,
  collectionPrompt: false
};

const authProps = {
  accountName: "",
  loggedIn: false,
  authPrompt: false
};

const routerProps = {
  pathname: "/test"
};

const staticProps = {
  pageName: "Collection"
};

const props = {
  ...dispatchProps,
  ...colProps,
  ...authProps,
  ...routerProps,
  ...staticProps
};

const colContainer = shallow(<CollectionContainer {...props} />);

describe("CollectionContainer", () => {
  test("renders correctly", () => {
    const component = renderer.create(<CollectionContainer {...props} />);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  test("onMount called on mount", () => {
    expect(dispatchProps.onMount).toBeCalled();
  });

  test("Header Component loaded", () => {
    expect(colContainer.find(HeaderComponent).length).toEqual(1);
  });

  test("Collection Component loaded", () => {
    expect(colContainer.find(CollectionComponent).length).toEqual(1);
  });
});
