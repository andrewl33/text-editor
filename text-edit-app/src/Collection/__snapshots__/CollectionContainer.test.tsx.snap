// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionContainer renders correctly 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
>
  <header>
    <div
      className="ui huge inverted attached menu"
      primary="true"
    >
      <a
        className="item"
        onClick={[Function]}
      >
        Text Edit
      </a>
      <div
        className="item"
        onClick={[Function]}
      >
        Collection
      </div>
      <div
        className="right menu"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            New
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            Lock
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <button
            className="ui primary button"
            onClick={[Function]}
            role="button"
          >
            Share
          </button>
        </div>
        <div
          className="item"
          onClick={[Function]}
        >
          <div>
            <div
              aria-expanded={false}
              className="ui button floating labeled dropdown icon"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              style={
                Object {
                  "backgroundColor": "#2185d0",
                  "color": "#fff",
                  "fontSize": 17,
                }
              }
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              >
                Account
              </div>
              <i
                aria-hidden="true"
                className="user icon"
                onClick={[Function]}
              />
              <div
                className="menu transition"
              >
                <div
                  className="item"
                  onClick={[Function]}
                  role="option"
                >
                  Log in
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="ui grid"
    style={
      Object {
        "height": "100%",
        "marginTop": 0,
      }
    }
  >
    <div
      className="thirteen wide column"
      style={
        Object {
          "paddingRight": 0,
          "paddingTop": 0,
        }
      }
    >
      <div>
        <table
          className="ui blue large celled inverted middle aligned table"
          divided="true"
          relaxed="true"
          style={
            Object {
              "borderRadius": 0,
              "color": "rgba(255,255,255,.9)",
            }
          }
        >
          <thead
            className=""
          >
            <tr
              className=""
            >
              <th
                className=""
                colSpan="2"
              >
                Files
              </th>
              <th
                className=""
              >
                Tags
              </th>
              <th
                className=""
                colSpan="2"
              >
                Dates
              </th>
              <th
                className=""
                colSpan="2"
              >
                Delete
              </th>
            </tr>
          </thead>
          <tbody
            className=""
          />
        </table>
        <div
          className="ui input"
        >
          <input
            onChange={[Function]}
            placeholder="File url..."
            type="text"
            value=""
          />
          <button
            className="ui button"
            onClick={[Function]}
            role="button"
            type="submit"
          >
            Add
          </button>
        </div>
      </div>
    </div>
    <div
      className="center aligned three wide column"
      style={
        Object {
          "backgroundColor": "#3180d4",
          "height": "100%",
          "paddingRight": "2em",
        }
      }
    >
      <div
        className="ui container"
      >
        <div
          className="ui list"
          role="list"
          style={
            Object {
              "height": "100%",
            }
          }
        >
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Name
            </div>
            <span
              className="name"
            >
              name
              <i
                aria-hidden="true"
                className="edit link icon"
                onClick={[Function]}
              />
            </span>
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Date Created
            </div>
            1-1-2001
          </div>
          <div
            className="item"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Private
            </div>
            No
          </div>
          <div
            className="item users"
            onClick={[Function]}
            role="listitem"
          >
            <div
              className="header"
            >
              Users
            </div>
            <span>
              No Users
            </span>
            
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
